# 🔰 First MileStone

# Start from Here
- Load all necessary libraries

In [ ]:
import pandas as pd #pandas for data manipulation
import json #json for reading json files
import numpy as np #numpy for numerical operations

## Try to use chunk for split the jsonl data first.
- as our file size is 14 GB and around 9 million datasize.
- use chunk size 500000 for better response

In [ ]:

chunk_size = 500000 # Number of rows to read at a time
records = [] # List to store the records
chunk_id = 0 # Initialize chunk ID

## Chunk the main files. called `datav17.jsonl`

In [ ]:
with open('datav17.jsonl', 'r', encoding='utf-8') as file: # Open the JSONL file for reading
    for i, line in enumerate(file): # Iterate over each line in the file
        try:
            records.append(json.loads(line))  # Attempt to parse the line as JSON and append it to the records list
        except json.JSONDecodeError:  # Handle cases where the line is not valid JSON
            continue  # Skip bad lines and proceed to the next iteration

        if (i + 1) % chunk_size == 0: # Check if the number of records has reached the chunk size
            df_chunk = pd.DataFrame(records) # Convert the records list to a DataFrame
            df_chunk.to_parquet(f"output_chunk_{chunk_id}.parquet") # Save the DataFrame to a Parquet file
            print(f"Saved chunk {chunk_id}")    # Print a message indicating the chunk has been saved
            chunk_id += 1
            records = []

    if records:
        df_chunk = pd.DataFrame(records) # Convert any remaining records to a DataFrame
        df_chunk.to_parquet(f"output_chunk_{chunk_id}.parquet") # Save any remaining records to a final chunk
        print(f"Saved chunk {chunk_id}") # Print a message indicating the final chunk has been saved


In [ ]:
# # Count how many lines (i.e., records) in the jsonl file
# with open('datav17.jsonl', 'r', encoding='utf-8') as f:
#     total_lines = sum(1 for _ in f)

# print(f"Total lines in file: {total_lines}")


In [ ]:
# import math
# total_lines = 6611132
# # chunk_size = 100000

# num_chunks = math.ceil(total_lines / chunk_size)
# print(f"Expected number of parquet files: {num_chunks}")


## Read all the parquest datasets

In [ ]:
#read the chunk
# df1 = pd.read_parquet("output_chunk_0.parquet")
# df2 = pd.read_parquet("output_chunk_1.parquet")
# df3 = pd.read_parquet("output_chunk_2.parquet")
# df4 = pd.read_parquet("output_chunk_3.parquet")
# df5 = pd.read_parquet("output_chunk_4.parquet")
# df6 = pd.read_parquet("output_chunk_5.parquet")
# df7 = pd.read_parquet("output_chunk_6.parquet")
# df8 = pd.read_parquet("output_chunk_7.parquet")
# df9 = pd.read_parquet("output_chunk_8.parquet")
# df10 = pd.read_parquet("output_chunk_9.parquet")
# df11 = pd.read_parquet("output_chunk_10.parquet")
# df12 = pd.read_parquet("output_chunk_11.parquet")
# df13 = pd.read_parquet("output_chunk_12.parquet")
# df14 = pd.read_parquet("output_chunk_13.parquet")


### Load all parquet files 
- converted some datatypes
- fix some issues related to data
- we have 14 parquet data, so we load different df for better response.

In [ ]:
import pandas as pd
import json
import numpy as np

df2 = pd.read_parquet("output_chunk_2.parquet")

# df2['first_author'] = df2['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df2['id'] = df2['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df2['name'] = df2['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df2['org'] = df2['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df2['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=10)

# Filter the dataframe for the last 10 years
filtered_df2 = df2[df2['year'] >= cutoff_date.year]
# filtered_df2.to_csv('output_chunk_0.csv', index=False)
filtered_df2.to_csv('output_chunk_2.csv', index=False)


In [ ]:
import pandas as pd
import json
import numpy as np

df3 = pd.read_parquet("output_chunk_3.parquet")

# df3['first_author'] = df3['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df3['id'] = df3['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df3['name'] = df3['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df3['org'] = df3['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df3['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=10)

# Filter the dataframe for the last 10 years
filtered_df3 = df3[df3['year'] >= cutoff_date.year]
# filtered_df3.to_csv('output_chunk_0.csv', index=False)
filtered_df3.to_csv('output_chunk_3.csv', index=False)

In [ ]:
import pandas as pd
import json
import numpy as np

df4 = pd.read_parquet("output_chunk_4.parquet")

# df4['first_author'] = df4['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df4['id'] = df4['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df4['name'] = df4['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df4['org'] = df4['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df4['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=10)

# Filter the dataframe for the last 10 years
filtered_df4 = df4[df4['year'] >= cutoff_date.year]
# filtered_df4.to_csv('output_chunk_0.csv', index=False)
filtered_df4.to_csv('output_chunk_4.csv', index=False)

In [ ]:
import pandas as pd
import json
import numpy as np

df5 = pd.read_parquet("output_chunk_5.parquet")

# df5['first_author'] = df5['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df5['id'] = df5['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df5['name'] = df5['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df5['org'] = df5['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df5['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=10)

# Filter the dataframe for the last 10 years
filtered_df5 = df5[df5['year'] >= cutoff_date.year]
# filtered_df5.to_csv('output_chunk_0.csv', index=False)
filtered_df5.to_csv('output_chunk_5.csv', index=False)

In [ ]:
import pandas as pd
import json
import numpy as np

df6 = pd.read_parquet("output_chunk_6.parquet")

# df6['first_author'] = df6['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df6['id'] = df6['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df6['name'] = df6['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df6['org'] = df6['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df6['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=10)

# Filter the dataframe for the last 10 years
filtered_df6 = df6[df6['year'] >= cutoff_date.year]
# filtered_df6.to_csv('output_chunk_0.csv', index=False)
filtered_df6.to_csv('output_chunk_6.csv', index=False)

In [ ]:
import pandas as pd
import json
import numpy as np

df7 = pd.read_parquet("output_chunk_7.parquet")

# df7['first_author'] = df7['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df7['id'] = df7['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df7['name'] = df7['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df7['org'] = df7['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df7['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=10)

# Filter the dataframe for the last 10 years
filtered_df7 = df7[df7['year'] >= cutoff_date.year]
# filtered_df7.to_csv('output_chunk_0.csv', index=False)
filtered_df7.to_csv('output_chunk_7.csv', index=False)

In [ ]:
import pandas as pd
import json
import numpy as np

df8 = pd.read_parquet("output_chunk_8.parquet")

# df8['first_author'] = df8['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df8['id'] = df8['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df8['name'] = df8['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df8['org'] = df8['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df8['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=10)

# Filter the dataframe for the last 10 years
filtered_df8 = df8[df8['year'] >= cutoff_date.year]
# filtered_df8.to_csv('output_chunk_0.csv', index=False)
filtered_df8.to_csv('output_chunk_8.csv', index=False)

In [ ]:
import pandas as pd
import json
import numpy as np

df9 = pd.read_parquet("output_chunk_9.parquet")

# df9['first_author'] = df9['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df9['id'] = df9['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df9['name'] = df9['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df9['org'] = df9['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df9['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=10)

# Filter the dataframe for the last 10 years
filtered_df9 = df9[df9['year'] >= cutoff_date.year]
# filtered_df9.to_csv('output_chunk_0.csv', index=False)
filtered_df9.to_csv('output_chunk_9.csv', index=False)

In [ ]:
import pandas as pd
import json
import numpy as np

df10 = pd.read_parquet("output_chunk_10.parquet")

# df10['first_author'] = df10['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df10['id'] = df10['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df10['name'] = df10['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df10['org'] = df10['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df10['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=10)

# Filter the dataframe for the last 10 years
filtered_df10 = df10[df10['year'] >= cutoff_date.year]
# filtered_df10.to_csv('output_chunk_0.csv', index=False)
filtered_df10.to_csv('output_chunk_10.csv', index=False)

In [ ]:
import pandas as pd
import json
import numpy as np

df11 = pd.read_parquet("output_chunk_11.parquet")

# df11['first_author'] = df11['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df11['id'] = df11['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df11['name'] = df11['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df11['org'] = df11['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df11['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=10)

# Filter the dataframe for the last 10 years
filtered_df11 = df11[df11['year'] >= cutoff_date.year]
# filtered_df11.to_csv('output_chunk_0.csv', index=False)
filtered_df11.to_csv('output_chunk_11.csv', index=False)

In [ ]:
import pandas as pd
import json
import numpy as np

df12 = pd.read_parquet("output_chunk_12.parquet")

# df12['first_author'] = df12['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df12['id'] = df12['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df12['name'] = df12['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df12['org'] = df12['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df12['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=12)

# Filter the dataframe for the last 12 years
filtered_df12 = df12[df12['year'] >= cutoff_date.year]
# filtered_df12.to_csv('output_chunk_0.csv', index=False)
filtered_df12.to_csv('output_chunk_12.csv', index=False)

In [ ]:
import pandas as pd
import json
import numpy as np

df13 = pd.read_parquet("output_chunk_13.parquet")

# df13['first_author'] = df13['authors'].apply(lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None)
import numpy as np
# df13['id'] = df13['authors'].apply(lambda x: x[0]['id'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
# no need the id
df13['name'] = df13['authors'].apply(lambda x: x[0]['name'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)
df13['org'] = df13['authors'].apply(lambda x: x[0]['org'] if isinstance(x, (list, np.ndarray)) and len(x) > 0 else None)


# df13['year'].value_counts().head()
cutoff_date = pd.to_datetime('today') - pd.DateOffset(years=13)

# Filter the dataframe for the last 13 years
filtered_df13 = df13[df13['year'] >= cutoff_date.year]
# filtered_df13.to_csv('output_chunk_0.csv', index=False)
filtered_df13.to_csv('output_chunk_13.csv', index=False)

## Read the all of our csv files with dataframe

In [ ]:
df0 = pd.read_csv('output_chunk_0.csv')
df1 = pd.read_csv('output_chunk_1.csv')
df2 = pd.read_csv('output_chunk_2.csv')
df3 = pd.read_csv('output_chunk_3.csv')
df4 = pd.read_csv('output_chunk_4.csv')
df5 = pd.read_csv('output_chunk_5.csv')
df6 = pd.read_csv('output_chunk_6.csv')
df7 = pd.read_csv('output_chunk_7.csv')
df8 = pd.read_csv('output_chunk_8.csv')
df9 = pd.read_csv('output_chunk_9.csv')
df10 = pd.read_csv('output_chunk_10.csv')
df11 = pd.read_csv('output_chunk_11.csv')
df12 = pd.read_csv('output_chunk_12.csv')
df13 = pd.read_csv('output_chunk_13.csv')

In [ ]:
#combine all the csv files into one
df_combined = pd.concat([df0, df1, df2, df3, df4, df5, df6, df7, df8, df9, df10, df11, df12, df13], ignore_index=True)
df_combined.to_csv('combined_output.csv', index=False)

In [ ]:
df = pd.read_csv('combined_output.csv')

In [ ]:
df['year'].value_counts().head(10)

year
2024    462618
2023    442008
2021    386661
2022    379730
2020    365566
2019    359917
2018    328635
2017    298205
2016    280918
2015    272715
Name: count, dtype: int64

In [ ]:
df['name'].value_counts().head(10)

name
Yang Liu     1262
Wei Wang     1157
Wei Zhang     868
Yang Li       855
Wei Li        818
Yu Zhang      792
Lei Wang      782
Yu Wang       763
Yang          748
Hao Wang      742
Name: count, dtype: int64

In [ ]:
df.isnull().sum()

id                  0
title               0
abstract       143294
keywords            0
year                0
authors             0
references          0
page_start    1180802
page_end      1220739
lang           150964
volume        1594429
issue         2541817
issn          1640596
isbn          3632271
doi            265074
url                 0
n_citation          0
venue           18624
doc_type         7368
name                5
org            532543
dtype: int64

In [ ]:
df['org'].value_counts().head(20)

org
Corresponding author.                                                        2961
Carnegie Mellon Univ, Pittsburgh, PA 15213 USA                               2839
Stanford Univ, Stanford, CA 94305 USA                                        1990
Georgia Inst Technol, Atlanta, GA 30332 USA                                  1949
Univ Calif Berkeley, Berkeley, CA 94720 USA                                  1742
Carnegie Mellon University                                                   1584
Tsinghua University                                                          1501
Univ Michigan, Ann Arbor, MI 48109 USA                                       1486
Tsinghua Univ, Beijing, Peoples R China                                      1483
Tsinghua Univ, Dept Elect Engn, Beijing 100084, Peoples R China              1439
Shanghai Jiao Tong Univ, Shanghai, Peoples R China                           1385
Nanyang Technol Univ, Sch Elect & Elect Engn, Singapore 639798, Singapore    1358
Tech Univ Mu

# 🔰 Second MileStone
## ✅ Tasks to Complete

**For this milestone, let's query the data for 4 things:**
- For each author What is the org for the last paper the author has published.
- Want to know for each author how many papers he was the first and in which org.
- For each author, how many authors they wrote a paper with him - total number
- For each author how many conferences and journals he published in -including their names (conferences and journals)
- Save this info in new columns - all in one csv


#### 🛠️ Pre-processing Requirement
Remove all authors with missing IDs before performing any of the above tasks.

##### Filtered Not_null:

In [ ]:
import pandas as pd
import ast

import json
# Read the dataset
df = pd.read_csv('combined_output.csv')


In [2]:
# data is correctly viewed
df.shape

(3603006, 19)

### Fixing the list of authors columns as this is the main pain points

In [5]:
# Function to safely parse the 'authors' column using json.loads
def safe_parse_json(x):
    try:
        # Try parsing the string as JSON (it will convert to a list/dictionary if valid)
        return json.loads(x) if isinstance(x, str) else x
    except json.JSONDecodeError:
        # If there's an error, return None or handle the error accordingly
        return None

# Apply the function to the 'authors' column
df['authors_parsed'] = df['authors'].apply(safe_parse_json)

# Check the parsed column
print(df['authors_parsed'].head())

0    [{'id': '560a8bf045cedb339712286b', 'name': 'E...
1    [{'id': '53f42c44dabfaee02ac50137', 'name': 'R...
2    [{'id': '5631093045cedb3399c5b9bb', 'name': 'K...
3    [{'id': '60e2c0766750f80c68e4ec61', 'name': 'Y...
4    [{'id': '53f4732fdabfaedd74e9c168', 'name': 'M...
Name: authors_parsed, dtype: object


In [6]:
df.shape

(3603006, 20)

##### Confirm all authors in authors_cleaned have 'id'

Clean Author Lists in Place (Keep Paper, Remove Only Authors Without ID)

In [7]:
df['authors_cleaned'] = df['authors_parsed'].apply(
    lambda authors: [a for a in authors if a.get('id')]
)

- Just checking

In [8]:
# Confirm all authors in authors_cleaned have 'id'
def all_authors_have_id(author_list):
    return all(a.get('id') for a in author_list)

# Filter the dataset
df_ready = df[df['authors_cleaned'].map(all_authors_have_id)]

# Quick review
print(f"Rows ready for analysis: {len(df_ready)}")
print(f"Sample row:\n{df_ready.iloc[0]}")


Rows ready for analysis: 3603006
Sample row:
id                                          53908b2120f70186a0db6fb6
title                                       Parsing and Collocations
abstract           Proper identification of collocations (and mor...
keywords           ['collocation', 'parsing', 'translation', 'dee...
year                                                            2022
authors            [{"id": "560a8bf045cedb339712286b", "name": "E...
references         ['5a73cbc317c44a0b3035e846', '5390985e20f70186...
page_start                                                       NaN
page_end                                                         NaN
lang                                                              en
volume                                                             5
issue                                                            NaN
issn                                                       2624-8212
isbn                                                      

In [11]:
df_ready.shape

(3603006, 21)

In [9]:
df_ready_clean = df_ready[df_ready['authors_parsed'].map(lambda authors: all(author.get('id') for author in authors))]


- `3603006` is without cleaned authors id
- `2633528` is with cleaned authors id

In [12]:
df_ready_clean.shape

(2633528, 21)

###  ✅ We successfully removed all authors with missing IDs

- as data has some values about 2027. 
- Don't know why but only 3 values I found
- I checked manually on web 2027 papers it's give me wrong papers year.

In [13]:
df_ready_clean.head(10)
# if possible save it to milestone 2 ready file

# your your need to be filter first
df_ready_clean_year =  df_ready_clean[df_ready_clean['year'] != 2027]
df_ready_clean_year.head()
# df_ready.to_csv('milestone2readyfile.csv', index=False)

id  \
0  53908b2120f70186a0db6fb6   
1  5390962020f70186a0df38ce   
2  5390994d20f70186a0e12262   
3  5390a8b220f70186a0e9c4e2   
4  5390a9a420f70186a0ea4cad   

                                               title  \
0                           Parsing and Collocations   
1  Integrating Text Planning and Linguistic Choic...   
2  The Aesthetics of Failure: "Post-Digital" Tend...   
3  Establishment of the Index System for Evaluati...   
4  PI Controller Design for Suppressing Exogenous...   

                                            abstract  \
0  Proper identification of collocations (and mor...   
1  Natural language generation is usually divided...   
2  Over the past decade, the Internet has helped ...   
3  An index system for the evaluation of the Brow...   
4  A novel approach is proposed to suppress bound...   

                                            keywords  year  \
0  ['collocation', 'parsing', 'translation', 'dee...  2022   
1                              ['Language Modeling']  2021   
2                    ['Digital Musical Instruments']  2017   
3  ['Brownfield redevelopment project', 'Evaluati...  2015   
4  ['linear system', 'exogenous disturbances', 'P...  2023   

                                             authors  \
0  [{"id": "560a8bf045cedb339712286b", "name": "E...   
1  [{"id": "53f42c44dabfaee02ac50137", "name": "R...   
2  [{"id": "5631093045cedb3399c5b9bb", "name": "K...   
3  [{"id": "60e2c0766750f80c68e4ec61", "name": "Y...   
4  [{"id": "53f4732fdabfaedd74e9c168", "name": "M...   

                                          references page_start page_end lang  \
0  ['5a73cbc317c44a0b3035e846', '5390985e20f70186...        NaN      NaN   en   
1                                                 []        NaN      NaN   en   
2                                                 []        547      554   en   
3  ['53e9b7eeb7602d97043a1bf3', '53e9a4fab7602d97...        173      182   en   
4  ['60435728d3485cfff1f6a31c', '61a75abf5244ab9d...        799      813   en   

   ... issue       issn isbn                            doi  \
0  ...   NaN  2624-8212  NaN       10.3389/frai.2022.765695   
1  ...     2        NaN  NaN        10.1162/089120100561656   
2  ...   NaN        NaN  NaN        10.1162/014892600559489   
3  ...   NaN  1364-8152  NaN  10.1016/j.envsoft.2015.09.012   
4  ...     8  0005-1179  NaN      10.1134/s0005117923080052   

                                                 url n_citation  \
0                                                 []         11   
1  ['https://dl.acm.org/doi/10.1162/0891201005616...          7   
2  ['https://dl.acm.org/doi/10.1162/0148926005594...        706   
3  ['https://openalex.org/W2344211555', ' https:/...         40   
4  ['https://link.springer.com/10.1134/S000511791...          7   

                                   venue    doc_type  \
0   Frontiers in Artificial Intelligence  Conference   
1              Computational Linguistics     Journal   
2                          Audio Culture     Journal   
3  Environmental Modelling &amp Software     Journal   
4          AUTOMATION AND REMOTE CONTROL     Journal   

                                      authors_parsed  \
0  [{'id': '560a8bf045cedb339712286b', 'name': 'E...   
1  [{'id': '53f42c44dabfaee02ac50137', 'name': 'R...   
2  [{'id': '5631093045cedb3399c5b9bb', 'name': 'K...   
3  [{'id': '60e2c0766750f80c68e4ec61', 'name': 'Y...   
4  [{'id': '53f4732fdabfaedd74e9c168', 'name': 'M...   

                                     authors_cleaned  
0  [{'id': '560a8bf045cedb339712286b', 'name': 'E...  
1  [{'id': '53f42c44dabfaee02ac50137', 'name': 'R...  
2  [{'id': '5631093045cedb3399c5b9bb', 'name': 'K...  
3  [{'id': '60e2c0766750f80c68e4ec61', 'name': 'Y...  
4  [{'id': '53f4732fdabfaedd74e9c168', 'name': 'M...  

[5 rows x 21 columns]

In [14]:
df_ready_clean_year.shape

(2633527, 21)

In [15]:
df_ready_clean_year.columns

Index(['id', 'title', 'abstract', 'keywords', 'year', 'authors', 'references',
       'page_start', 'page_end', 'lang', 'volume', 'issue', 'issn', 'isbn',
       'doi', 'url', 'n_citation', 'venue', 'doc_type', 'authors_parsed',
       'authors_cleaned'],
      dtype='object')

# ✅ 1. Last Paper Published by Each Author (with org)
- For each author: return last paper ID, title, name, author ID, and org.

In [16]:
# Explode to author-level records
author_papers = []

for idx, row in df_ready_clean_year.iterrows():
    for author in row['authors_cleaned']:
        author_papers.append({
            'author_id': author.get('id'),
            'author_name': author.get('name'),
            'author_org': author.get('org'),
            'paper_id': row.get('id'),
            'title': row.get('title'),
            'year': row.get('year')
        })

author_df = pd.DataFrame(author_papers)

# Sort and get last paper per author by year
last_papers = (
    author_df.sort_values(by='year', ascending=False)
.drop_duplicates(subset=['author_id'], keep='first')
)

print("📄 Last paper for each author:")
print(last_papers.head())


📄 Last paper for each author:
                       author_id     author_name  \
408098  64c137889fbb79c7b8cbf670  Chunxiao Zhang   
715996  62e4bf35d9f204418d71f3cc      Xuting Wei   
715993  6519f517768b11dc72ad8470   Chengjun Kang   
715994  62e8b143d9f20422f0bc912e  Yongsheng Qian   
715995  53f463c5dabfaedf4363b87f     Junwei Zeng   

                                               author_org  \
408098  Shandong Normal Univ, Sch Math Sci, Jinan 2500...   
715996  Lanzhou Jiaotong Univ, Sch Traff & Transportat...   
715993  Lanzhou Jiaotong Univ, Sch Traff & Transportat...   
715994  Lanzhou Jiaotong Univ, Sch Traff & Transportat...   
715995  Lanzhou Jiaotong Univ, Sch Traff & Transportat...   

                        paper_id  \
408098  644744fa71ac66d2cbf9b199   
715996  6747ab8cae8580e7ff69c5e0   
715993  6747ab8cae8580e7ff69c5e0   
715994  6747ab8cae8580e7ff69c5e0   
715995  6747ab8cae8580e7ff69c5e0   

                                                    title  year  
408098 

In [17]:
last_papers.head()

author_id     author_name  \
408098  64c137889fbb79c7b8cbf670  Chunxiao Zhang   
715996  62e4bf35d9f204418d71f3cc      Xuting Wei   
715993  6519f517768b11dc72ad8470   Chengjun Kang   
715994  62e8b143d9f20422f0bc912e  Yongsheng Qian   
715995  53f463c5dabfaedf4363b87f     Junwei Zeng   

                                               author_org  \
408098  Shandong Normal Univ, Sch Math Sci, Jinan 2500...   
715996  Lanzhou Jiaotong Univ, Sch Traff & Transportat...   
715993  Lanzhou Jiaotong Univ, Sch Traff & Transportat...   
715994  Lanzhou Jiaotong Univ, Sch Traff & Transportat...   
715995  Lanzhou Jiaotong Univ, Sch Traff & Transportat...   

                        paper_id  \
408098  644744fa71ac66d2cbf9b199   
715996  6747ab8cae8580e7ff69c5e0   
715993  6747ab8cae8580e7ff69c5e0   
715994  6747ab8cae8580e7ff69c5e0   
715995  6747ab8cae8580e7ff69c5e0   

                                                    title  year  
408098  Uniform Convergence of Finite Element Method o...  2025  
715996  A Double Flux Control Strategy for an Improved...  2025  
715993  A Double Flux Control Strategy for an Improved...  2025  
715994  A Double Flux Control Strategy for an Improved...  2025  
715995  A Double Flux Control Strategy for an Improved...  2025

# To save this file
If run first time don't do this again

## I will used this data for ans the question `3`
- df_ready_clean_year.to_csv('cleaned_for2milestone.csv', index=False) 

In [15]:
# cleaned_last_paper.to_csv('last_paper.csv', index=False)
# you can also filter this before save (milestone2readyfile.csv)
# df_ready_clean_year.to_csv('cleaned_for2milestone.csv', index=False) 

1. `Exclude in a separate CSV file, all the authors’ names in one column, and the other column their ID (multiple csv if it is too long)`- you can add column or remove one

In [ ]:
all_authorsName_with_id = last_papers[['author_name','author_id','author_org']]
all_authorsName_with_id.head()


author_name                 author_id  \
408098  Chunxiao Zhang  64c137889fbb79c7b8cbf670   
715996      Xuting Wei  62e4bf35d9f204418d71f3cc   
715993   Chengjun Kang  6519f517768b11dc72ad8470   
715994  Yongsheng Qian  62e8b143d9f20422f0bc912e   
715995     Junwei Zeng  53f463c5dabfaedf4363b87f   

                                               author_org  
408098  Shandong Normal Univ, Sch Math Sci, Jinan 2500...  
715996  Lanzhou Jiaotong Univ, Sch Traff & Transportat...  
715993  Lanzhou Jiaotong Univ, Sch Traff & Transportat...  
715994  Lanzhou Jiaotong Univ, Sch Traff & Transportat...  
715995  Lanzhou Jiaotong Univ, Sch Traff & Transportat...

Run this for save this into cvs

In [ ]:
# all_authorsName_with_id.to_csv('all_authorsName_with_id.csv', index=False) 

 = last_papers.head()a
 what i have to do? all authors and their id. export last_papers has the actual values after removed missing id


💥✔ This saved files is worked for mile stone 2

# ✅ 2. Count of Papers Where Author is First Author (+ Org)
- Return: paper ID, title, author ID, name, org (only where author is first in list)

In [20]:
first_author_rows = []

for idx, row in df_ready_clean_year.iterrows():
    if row['authors_cleaned']:
        first_author = row['authors_cleaned'][0]
        first_author_rows.append({
            'author_id': first_author.get('id'),
            'author_name': first_author.get('name'),
            'author_org': first_author.get('org'),
            'paper_id': row.get('id'),
            'title': row.get('title'),
        })

first_author_df = pd.DataFrame(first_author_rows)

# Count of first-authored papers per author
first_author_counts = first_author_df['author_id'].value_counts().reset_index()
first_author_counts.columns = ['author_id', 'first_author_paper_count']

# Merge back for detailed info
first_author_details = pd.merge(
    first_author_df, first_author_counts, on='author_id', how='left'
).drop_duplicates()

print("📌 First-author papers summary:")
print(first_author_details.head())



📌 First-author papers summary:
                  author_id       author_name  \
0  560a8bf045cedb339712286b       Eric Wehrli   
1  53f42c44dabfaee02ac50137        R Rubinoff   
2  5631093045cedb3399c5b9bb         K Cascone   
3  60e2c0766750f80c68e4ec61        Yuming Zhu   
4  53f4732fdabfaedd74e9c168  M. V. Khlebnikov   

                                          author_org  \
0              LATL,University of Geneva,Switzerland   
1  AnswerLog Inc, 1111 19 St NW,Suite 600, Washin...   
2              Anechoicmedia, Pacifica, CA 94044 USA   
3  Northwestern Polytech Univ, Sch Management, Xi...   
4  Trapeznikov Institute of Control Sciences, Rus...   

                   paper_id  \
0  53908b2120f70186a0db6fb6   
1  5390962020f70186a0df38ce   
2  5390994d20f70186a0e12262   
3  5390a8b220f70186a0e9c4e2   
4  5390a9a420f70186a0ea4cad   

                                               title  first_author_paper_count  
0                           Parsing and Collocations                 

- You can save this dataframe as csv also

In [ ]:
# first_author_details.to_csv('first_authors_org_count.csv', index=False)

#### Now sort for those who are top authors
- this is not final result

In [21]:
first_author_details.sort_values(by='first_author_paper_count', ascending=False).head(20)

author_id        author_name  \
725059   53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
1858630  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
1414285  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
768553   53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
380279   53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
2412523  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
444309   53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
2046801  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
1885269  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
983116   53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
2496267  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
1750437  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
1346237  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
2619876  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
752181   53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
1794777  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
1960487  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
994678   53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
458787   53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   
1620782  53f4388ddabfaee0d9b78cb9  Alfredo Cuzzocrea   

                                                author_org  \
725059                                                None   
1858630               ICAR CNR, Arcavacata Di Rende, Italy   
1414285                       Univ Trieste, Trieste, Italy   
768553                                                None   
380279      Univ Trieste, DIA Dept, I-34127 Trieste, Italy   
2412523                           ICAR CNR, Cosenza, Italy   
444309                        Univ Trieste, Trieste, Italy   
2046801                               ICAR-CNR,Rende,Italy   
1885269                       Univ Trieste, Trieste, Italy   
983116   DIA Department,University of Trieste,Trieste,I...   
2496267             Univ Trieste, DIA Dept, Trieste, Italy   
1750437   Dept. DIA, University of Trieste, Trieste, Italy   
1346237                                               None   
2619876  DISPES Department, University of Calabria, Ren...   
752181           University of Trieste and ICAR-CNR, Italy   
1794777  DIA Department,University of Trieste and ICAR-...   
1960487  Univ Calabria, iDEA Lab, Arcavacata Di Rende, ...   
994678   DIA Department,University of Trieste and ICAR-...   
458787                              University of Calabria   
1620782                                               None   

                         paper_id  \
725059   6523b74a939a5f40821b93d2   
1858630  573697d96e3b12023e6b7ef3   
1414285  5a73cb6a17c44a0b303590ea   
768553   6135f66191e011c9de642454   
380279   573695d26e3b12023e4eaa3b   
2412523  573697ae6e3b12023e691471   
444309   5b8c9ef817c44af36f8b0c4e   
2046801  5b67b47917c44aac1c863464   
1885269  5a4aef4b17c44a2190f769f9   
983116   573698446e3b12023e70d3ab   
2496267  599c7c3d601a182cd2797a8a   
1750437  5b67b46417c44aac1c8619b3   
1346237  614325d75244ab9dcb297bd4   
2619876  5e29721f3a55ac6b5c088713   
752181   57d063dcac44367354293dbb   
1794777  599c7be0601a182cd276e779   
1960487  63d7a22690e50fcafd6d618c   
994678   64e8f5213fda6d7f0643812e   
458787   5f27e03591e0118126340102   
1620782  5a73cb4217c44a0b30355f6b   

                                                     title  \
725059   Graph Representation Learning for Complex Secu...   
1858630  Runtime Anomaly Detection in Embedded Systems ...   
1414285  Data Masking Techniques for NoSQL Database Sec...   
768553   A Novel Framework for Supporting Mobile Object...   
380279   An Effective and Efficient MapReduce Algorithm...   
2412523  Spatio-temporal Analysis of Greenhouse Gas Dat...   
444309   Towards Lambda-Based Near Real-Time OLAP over ...   
2046801  An Innovative Architecture for Supporting Cybe...   
1885269  Privacy-Preserving Big Data Stream Mining: Opp...   
983116   A Multi-view Learning Approach to the Discover...   
2496267  A Novel Approach For Supporting Approximate Re...   
1750437  A Predictive 

- You can simply see only author_name, author_org only with first authors total count 

In [24]:
first_author_details[['author_name', 'author_org', 'first_author_paper_count']].sort_values(by='first_author_paper_count', ascending=False).head(20)


author_name                                         author_org  \
725059   Alfredo Cuzzocrea                                               None   
1858630  Alfredo Cuzzocrea               ICAR CNR, Arcavacata Di Rende, Italy   
1414285  Alfredo Cuzzocrea                       Univ Trieste, Trieste, Italy   
768553   Alfredo Cuzzocrea                                               None   
380279   Alfredo Cuzzocrea     Univ Trieste, DIA Dept, I-34127 Trieste, Italy   
2412523  Alfredo Cuzzocrea                           ICAR CNR, Cosenza, Italy   
444309   Alfredo Cuzzocrea                       Univ Trieste, Trieste, Italy   
2046801  Alfredo Cuzzocrea                               ICAR-CNR,Rende,Italy   
1885269  Alfredo Cuzzocrea                       Univ Trieste, Trieste, Italy   
983116   Alfredo Cuzzocrea  DIA Department,University of Trieste,Trieste,I...   
2496267  Alfredo Cuzzocrea             Univ Trieste, DIA Dept, Trieste, Italy   
1750437  Alfredo Cuzzocrea   Dept. DIA, University of Trieste, Trieste, Italy   
1346237  Alfredo Cuzzocrea                                               None   
2619876  Alfredo Cuzzocrea  DISPES Department, University of Calabria, Ren...   
752181   Alfredo Cuzzocrea          University of Trieste and ICAR-CNR, Italy   
1794777  Alfredo Cuzzocrea  DIA Department,University of Trieste and ICAR-...   
1960487  Alfredo Cuzzocrea  Univ Calabria, iDEA Lab, Arcavacata Di Rende, ...   
994678   Alfredo Cuzzocrea  DIA Department,University of Trieste and ICAR-...   
458787   Alfredo Cuzzocrea                             University of Calabria   
1620782  Alfredo Cuzzocrea                                               None   

         first_author_paper_count  
725059                        199  
1858630                       199  
1414285                       199  
768553                        199  
380279                        199  
2412523                       199  
444309                        199  
2046801                       199  
1885269                       199  
983116                        199  
2496267                       199  
1750437                       199  
1346237                       199  
2619876                       199  
752181                        199  
1794777                       199  
1960487                       199  
994678                        199  
458787                        199  
1620782                       199

- see top 5 values

In [27]:
first_author_details.head()

author_id       author_name  \
0  560a8bf045cedb339712286b       Eric Wehrli   
1  53f42c44dabfaee02ac50137        R Rubinoff   
2  5631093045cedb3399c5b9bb         K Cascone   
3  60e2c0766750f80c68e4ec61        Yuming Zhu   
4  53f4732fdabfaedd74e9c168  M. V. Khlebnikov   

                                          author_org  \
0              LATL,University of Geneva,Switzerland   
1  AnswerLog Inc, 1111 19 St NW,Suite 600, Washin...   
2              Anechoicmedia, Pacifica, CA 94044 USA   
3  Northwestern Polytech Univ, Sch Management, Xi...   
4  Trapeznikov Institute of Control Sciences, Rus...   

                   paper_id  \
0  53908b2120f70186a0db6fb6   
1  5390962020f70186a0df38ce   
2  5390994d20f70186a0e12262   
3  5390a8b220f70186a0e9c4e2   
4  5390a9a420f70186a0ea4cad   

                                               title  first_author_paper_count  
0                           Parsing and Collocations                         2  
1  Integrating Text Planning and Linguistic Choic...                         1  
2  The Aesthetics of Failure: "Post-Digital" Tend...                         1  
3  Establishment of the Index System for Evaluati...                         1  
4  PI Controller Design for Suppressing Exogenous...                        16

# ✔ Done 

# ✅ 3. Co-authors per Author (Comma-separated)
- For each author: list co-author name+id combinations, and count of co-authors.
- paper id has some duplicates values we removed it
- need to filter unique one first

Previously I saved this data.

In [1]:
from collections import defaultdict
import pandas as pd
cleaned_data =pd.read_csv('cleaned_for2milestone.csv')

In [2]:
cleaned_data.head(20)
# still we have the right data

id  \
0   53908b2120f70186a0db6fb6   
1   5390962020f70186a0df38ce   
2   5390994d20f70186a0e12262   
3   5390a8b220f70186a0e9c4e2   
4   5390a9a420f70186a0ea4cad   
5   5390ac1720f70186a0eb282b   
6   5390ad5620f70186a0ebd995   
7   5390aefc20f70186a0ece31f   
8   5390b00c20f70186a0ed5f49   
9   5390b78a20f70186a0f23cd6   
10  53e997c6b7602d9701fb70f8   
11  53e9993fb7602d970217bd65   
12  53e999b5b7602d97021fb749   
13  53e99a7fb7602d97022f056c   
14  53e9a3c0b7602d9702ccea39   
15  53e9a4d6b7602d9702df5ce3   
16  53e9a894b7602d97031e3638   
17  53e9a8bdb7602d970320aa8b   
18  53e9aa9cb7602d9703418595   
19  53e9ade2b7602d97037eea8f   

                                                title  \
0                            Parsing and Collocations   
1   Integrating Text Planning and Linguistic Choic...   
2   The Aesthetics of Failure: "Post-Digital" Tend...   
3   Establishment of the Index System for Evaluati...   
4   PI Controller Design for Suppressing Exogenous...   
5   The Price of Anarchy in Bilateral Network Form...   
6   Proof-theoretic Semantics for a Natural Langua...   
7   On the V-line Radon Transform and Its Imaging ...   
8                            Letter from the Editors.   
9   An Advanced Image Analysis Tool for the Quanti...   
10                                  Randomer Forests.   
11  Remark on Algorithm 897: VTDIRECT95: Serial an...   
12  On the Improvement of Remote Authentication Sc...   
13  Low Dimensional Trajectory Hypothesis is True:...   
14  Analyzing Dimension Mappings and Properties in...   
15  Design of Fixed-Lag Smoother Using Covariance ...   
16  Memoized Zipper-Based Attribute Grammars and T...   
17  Large-deviation Properties of the Largest Bico...   
18  Dirac Structures on Hilbert Spaces and Boundar...   
19  Retraction Note To: Application of Genetic Pro...   

                                             abstract  \
0   Proper identification of collocations (and mor...   
1   Natural language generation is usually divided...   
2   Over the past decade, the Internet has helped ...   
3   An index system for the evaluation of the Brow...   
4   A novel approach is proposed to suppress bound...   
5   We study network formation with n players and ...   
6   The paper has two parts: 1. A brief exposition...   
7   Radon transforms defined on smooth curves are ...   
8                                                 NaN   
9   The paper presents an advanced image analysis ...   
10  Decision forests, including Random Forests and...   
11  The Fortran95 code VTDIRECT95, based on the or...   
12  In 2005, Sun et al's proposed a user-friendly ...   
13  Deep neural networks (DNNs) usually contain ma...   
14  The information inside the Data Warehouse (DW)...   
15  A recursive fixed-point smoother and filter, f...   
16  Attribute grammars are a suitable formalism to...   
17  Distributions of the size of the largest compo...   
18  The aim of this work is to show how the Dirac ...   
19  Soft computing modeling of strength enhancemen...   

                                             keywords  year  \
0   ['collocation', 'parsing', 'translation', 'dee...  2022   
1                               ['Language Modeling']  2021   
2                     ['Digital Musical Instruments']  2017   
3   ['Brownfield redevelopment project', 'Evaluati...  2015   
4   ['linear system', 'exogenous disturbances', 'P...  2023   
5   ['Network formation', 'Bilateral link formatio...  2016   
6   ['Proof-theoretic semantics', 'Semantics of na...  2019   
7   ['Radon transforms', 'image reconstruction', '...  2015   
8                                                  []  2021   
9   ['Breast cancer', 'Pathology quantification', ...  2015   
10                                                 []  2015   
11  ['Algorithms', 'Design', 'Documentation', 'DIR...  2015   
12  ['remote authentication', 'privileged insider ...  2016   
13  ['Training', 'Trajectory', 'Neural networks', ...  2023   
14  ['Semant

In [3]:
import ast

# If authors_cleaned is a list of stringified dicts (check a few rows to confirm), fix it:
cleaned_data['authors_cleaned'] = cleaned_data['authors_cleaned'].apply(
    lambda x: ast.literal_eval(x) if isinstance(x, str) else x
)


### Try to drop duplicates ID first, as it aftter the final count of co authors

- We will use this cleaned_data for ans the query `4`

In [4]:
unique_papers_id = cleaned_data.drop_duplicates(subset='id')

# This is our actual values as paper `id` is a unique one for per paper. 
- so our total unique values for last 10 years is 2633527

In [5]:
unique_papers_id.shape

(2633527, 21)

In [6]:

# Step 2: Build the summary per paper
records = []

for row in unique_papers_id.itertuples():
    paper_id = row.id
    authors = row.authors_cleaned
    
    if not authors or len(authors) < 2:
        continue  # skip if no coauthors

    first_author = authors[0]
    first_author_id = first_author.get('id')
    first_author_name = first_author.get('name')

    coauthor_count = len([a for a in authors if a.get('id') != first_author_id])

    records.append({
        'paper_id': paper_id,
        'first_author_id': first_author_id,
        'first_author_name': first_author_name,
        'total_coauthors': coauthor_count
    })

# Step 3: Create DataFrame
coauthor_df = pd.DataFrame(records)

# Step 4: Preview and export
print(coauthor_df.sort_values(by='total_coauthors', ascending=False).head())
coauthor_df.to_csv("coauthor_summary_by_first_author.csv", index=False)


                         paper_id           first_author_id first_author_name  \
19872    64a3a1fcd68f896efa46704e  53f4461adabfaeee229f660f         R. Abbasi   
370306   63195d8890e50fcafde7e682  53f4461adabfaeee229f660f         R. Abbasi   
2173379  60129be991e011b3f52c1586  53f4461adabfaeee229f660f         R. Abbasi   
45081    5fabc05691e0110281fdabe4  6402cd834fc97c2f7417853b       M. Pontinen   
134967   5f0dc9a89fced0a24b1220d2  53f42d4ddabfaeb1a7b89ef8     Matej Kristan   

         total_coauthors  
19872                379  
370306               379  
2173379              364  
45081                116  
134967               103  


- to check how many less than equal 15

In [7]:
coauthor_less_15 = coauthor_df[coauthor_df['total_coauthors'] <= 15]
coauthor_less_15.head()


paper_id           first_author_id   first_author_name  \
0  5390a8b220f70186a0e9c4e2  60e2c0766750f80c68e4ec61          Yuming Zhu   
1  5390aefc20f70186a0ece31f  53f45d87dabfaee0d9c0dc6a        M. Morvidone   
2  5390b00c20f70186a0ed5f49  53f43274dabfaedf435570fd  M Michael Sathekge   
3  5390b78a20f70186a0f23cd6  53f430b7dabfaee2a1ca7057   Theodosios Goudas   
4  53e997c6b7602d9701fb70f8  637872d4d49364c9ac9815aa     Tyler M. Tomita   

   total_coauthors  
0                2  
1                3  
2                1  
3                1  
4               10

- Count less than equal 15

In [8]:
coauthor_less_15.shape

(2398209, 4)

- to check how many greater than 15

In [9]:
coauthor_more_15 = coauthor_df[coauthor_df['total_coauthors'] >= 15]
coauthor_more_15.head()

paper_id           first_author_id  \
161   56278673ad865a7dd91ae84e  53f455e2dabfaedd74e2bf38   
1695  5b67b4a717c44aac1c865bcc  54489c7ddabfae87b7e4f3eb   
1792  5bdc313a17c44a1f58a03cf3  5606f0a145ce1e595ede07b8   
1915  5c136da1da56295a08aada18  53f42d93dabfaeb2acfece9f   
2297  5cdfe4f6e1cd8edcef96920b  53f44205dabfaeb22f4b108b   

              first_author_name  total_coauthors  
161            Robert Petryszak               23  
1695              Dumitru Roman               16  
1792   Diofantos G. Hadjimitsis               17  
1915  Dionysios A. Antonopoulos               28  
2297         Takeharu Goji Etoh               15

- total number greater than 15

In [10]:
coauthor_more_15.shape

(3642, 4)

To save the Queary 3

In [ ]:
# coauthor_df.to_csv("coauthors_summary.csv", index=False)

In [11]:
coauthor_df.shape

(2401074, 4)

# ✔ Done

# ✅ 4. Conference and Journal Counts per Author (With Names)
- For each author: list names of conferences and journals, and total counts.

In [12]:
cleaned_data.head()

id  \
0  53908b2120f70186a0db6fb6   
1  5390962020f70186a0df38ce   
2  5390994d20f70186a0e12262   
3  5390a8b220f70186a0e9c4e2   
4  5390a9a420f70186a0ea4cad   

                                               title  \
0                           Parsing and Collocations   
1  Integrating Text Planning and Linguistic Choic...   
2  The Aesthetics of Failure: "Post-Digital" Tend...   
3  Establishment of the Index System for Evaluati...   
4  PI Controller Design for Suppressing Exogenous...   

                                            abstract  \
0  Proper identification of collocations (and mor...   
1  Natural language generation is usually divided...   
2  Over the past decade, the Internet has helped ...   
3  An index system for the evaluation of the Brow...   
4  A novel approach is proposed to suppress bound...   

                                            keywords  year  \
0  ['collocation', 'parsing', 'translation', 'dee...  2022   
1                              ['Language Modeling']  2021   
2                    ['Digital Musical Instruments']  2017   
3  ['Brownfield redevelopment project', 'Evaluati...  2015   
4  ['linear system', 'exogenous disturbances', 'P...  2023   

                                             authors  \
0  [{"id": "560a8bf045cedb339712286b", "name": "E...   
1  [{"id": "53f42c44dabfaee02ac50137", "name": "R...   
2  [{"id": "5631093045cedb3399c5b9bb", "name": "K...   
3  [{"id": "60e2c0766750f80c68e4ec61", "name": "Y...   
4  [{"id": "53f4732fdabfaedd74e9c168", "name": "M...   

                                          references page_start page_end lang  \
0  ['5a73cbc317c44a0b3035e846', '5390985e20f70186...        NaN      NaN   en   
1                                                 []        NaN      NaN   en   
2                                                 []        547      554   en   
3  ['53e9b7eeb7602d97043a1bf3', '53e9a4fab7602d97...        173      182   en   
4  ['60435728d3485cfff1f6a31c', '61a75abf5244ab9d...        799      813   en   

   ... issue       issn isbn                            doi  \
0  ...   NaN  2624-8212  NaN       10.3389/frai.2022.765695   
1  ...     2        NaN  NaN        10.1162/089120100561656   
2  ...   NaN        NaN  NaN        10.1162/014892600559489   
3  ...   NaN  1364-8152  NaN  10.1016/j.envsoft.2015.09.012   
4  ...     8  0005-1179  NaN      10.1134/s0005117923080052   

                                                 url n_citation  \
0                                                 []         11   
1  ['https://dl.acm.org/doi/10.1162/0891201005616...          7   
2  ['https://dl.acm.org/doi/10.1162/0148926005594...        706   
3  ['https://openalex.org/W2344211555', ' https:/...         40   
4  ['https://link.springer.com/10.1134/S000511791...          7   

                                   venue    doc_type  \
0   Frontiers in Artificial Intelligence  Conference   
1              Computational Linguistics     Journal   
2                          Audio Culture     Journal   
3  Environmental Modelling &amp Software     Journal   
4          AUTOMATION AND REMOTE CONTROL     Journal   

                                      authors_parsed  \
0  [{'id': '560a8bf045cedb339712286b', 'name': 'E...   
1  [{'id': '53f42c44dabfaee02ac50137', 'name': 'R...   
2  [{'id': '5631093045cedb3399c5b9bb', 'name': 'K...   
3  [{'id': '60e2c0766750f80c68e4ec61', 'name': 'Y...   
4  [{'id': '53f4732fdabfaedd74e9c168', 'name': 'M...   

                                     authors_cleaned  
0  [{'id': '560a8bf045cedb339712286b', 'name': 'E...  
1  [{'id': '53f42c44dabfaee02ac50137', 'name': 'R...  
2  [{'id': '5631093045cedb3399c5b9bb', 'name': 'K...  
3  [{'id': '60e2c0766750f80c68e4ec61', 'name': 'Y...  
4  [{'id': '53f4732fdabfaedd74e9c168', 'name': 'M...  

[5 rows x 21 columns]

In [13]:
unique_papers_id['id'].value_counts(ascending=False).head()

id
53908b2120f70186a0db6fb6    1
58437713ac44360f1082d34d    1
57d063e8ac44367354294d44    1
57d063e8ac44367354294f6f    1
57d063e8ac44367354294f9d    1
Name: count, dtype: int64

In [14]:

# Step 2: Initialize tracking dictionary
author_type_counts = defaultdict(lambda: {
    'name': '',
    'total_conferences': 0,
    'total_journals': 0
})
# Step 3: Process each unique paper
for row in unique_papers_id.itertuples():
    paper_id = getattr(row, 'id', None)
    authors = getattr(row, 'authors_cleaned', [])
    doc_type = str(getattr(row, 'doc_type', '')).lower()  # safe conversion

    if not authors or not paper_id or doc_type not in ['conference', 'journal']:
        continue

    # First author only
    first_author = authors[0]
    aid = first_author.get('id')
    name = first_author.get('name', '')

    if not aid:
        continue

    # Count doc_type per author
    if doc_type == 'conference':
        author_type_counts[aid]['total_conferences'] += 1
    elif doc_type == 'journal':
        author_type_counts[aid]['total_journals'] += 1

    author_type_counts[aid]['name'] = name

# Step 4: Build final DataFrame
first_author_df = pd.DataFrame([
    {
        'first_author_id': aid,
        'first_author_name': data['name'],
        'total_conferences': data['total_conferences'],
        'total_journals': data['total_journals']
    }
    for aid, data in author_type_counts.items()
])

# Display preview
print("✅ Final First Author Publication Summary:")
print(first_author_df.head())

# Optional: save to file
# first_author_df.to_csv("first_author_publication_summary.csv", index=False)


✅ Final First Author Publication Summary:
            first_author_id first_author_name  total_conferences  \
0  560a8bf045cedb339712286b       Eric Wehrli                  2   
1  53f42c44dabfaee02ac50137        R Rubinoff                  0   
2  5631093045cedb3399c5b9bb         K Cascone                  0   
3  60e2c0766750f80c68e4ec61        Yuming Zhu                  0   
4  53f4732fdabfaedd74e9c168  M. V. Khlebnikov                  4   

   total_journals  
0               0  
1               1  
2               1  
3               1  
4              12  


In [15]:
first_author_df.sort_values(by=['total_conferences', 'total_journals'], ascending=[False, False])


first_author_id     first_author_name  total_conferences  \
15540   53f4388ddabfaee0d9b78cb9     Alfredo Cuzzocrea                168   
24841   53f44b96dabfaee4dc7e6850   Huber Nieto-Chaupis                 82   
15979   53f42af4dabfaeb2acfc22f3  Alexander V. Smirnov                 73   
32267   53f43e27dabfaefedbb006d2            Dirk Beyer                 70   
9167    53f43f73dabfaedce5569454           Flora Amato                 68   
...                          ...                   ...                ...   
990642  638080e2fc451b2d602a8a08          Wonkwang Lee                  0   
990643  6525f14c55b3f8ac46fc7b34           Igor Pusnik                  0   
990644  6403e1d7adc7183bcb7eecb1        Naveed Taimoor                  0   
990645  632952c130241ae3e5b078c8        S. Poonguzhali                  0   
990646  6402ff907691d561fb05e658  Shehu Mohammed Yusuf                  0   

        total_journals  
15540               31  
24841                0  
15979                5  
32267               16  
9167                31  
...                ...  
990642               1  
990643               1  
990644               1  
990645               1  
990646               1  

[990647 rows x 4 columns]

# ✔ Done